### ISSUE TITLE
Add `session.interpreter` @property?

### ISSUE THREAD

**Issue by dhermes** [theacodes/nox#125]
This would redirect to `self.virtualenv.interpreter`. As a follow-up to this, I think there is another question:

- Should `Virtualenv._resolve_interpreter` be allowed to mutate the value of `self.interpreter`. Modifying the value makes it hard to do checks like `if session.interpreter == "foo"` in a user's `noxfile.py`, since the value the user passed in may change
- Should `Virtualenv._resolve_interpreter` be cached forever after the first usage? I think this answer is a resounding yes?

**Comment by dhermes** [comment]
To add flame to the fire here, the session names correspond to the original passed in value of `interpreter` (i.e. `func.python`) but `session.virtualenv.interpreter` may be changed due to the side effects of `_resolve_interpreter`

**Comment by theacodes** [comment]
What would this be used for?

**Comment by dhermes** [comment]
As an [example][1]:

```python
@nox.session(py=ALL_INTERPRETERS)
def unit(session):
    interpreter = session.virtualenv.interpreter
    if interpreter == PYPY:
        local_deps = pypy_setup(BASE_DEPS, session)
    else:
        local_deps = BASE_DEPS + (DEPS["scipy"],)
    if interpreter == PYPY or "2.7" in interpreter:
        local_deps += (DEPS["mock"],)
```

The `scipy` and `mock` deps could be handled with [environment markers][2] (e.g. `; python_version == '2.7'` and `; platform_python_implementation == 'PyPy'`). However the `pypy_setup` helper actually adds the `--no-index` flag to `pip install` and points to pre-compiled wheels via `--find-links ${WHEELHOUSE}`. This can't really be done in a `requirements.txt` since `${WHEELHOUSE}` is determined at runtime.

[1]: https://github.com/dhermes/bezier/blob/nox-update/noxfile.py#L151-L159
[2]: https://www.python.org/dev/peps/pep-0496/

**Comment by dhermes** [comment]
I sent #127 for the second part (i.e. not modifying `self.interpreter` on instances of `Virtualenv`)

**Comment by theacodes** [comment]
@dhermes is anything left to do here?

**Comment by dhermes** [comment]
Yes. The prerequisite to adding the `@property` was that the value doesn't get mutated (done in #127). Now we can add it and it should be straightforward.

**Comment by theacodes** [comment]
Cool. :)

**Comment by dhermes** [comment]
@theacodes your commit didn't add this feature

**Comment by theacodes** [comment]
@dhermes what am I missing? Pulling from your example above, this works:

```python
@nox.session(py=ALL_INTERPRETERS)
def unit(session):
    if session.python == PYPY:
        local_deps = pypy_setup(BASE_DEPS, session)
    else:
        local_deps = BASE_DEPS + (DEPS["scipy"],)
    if session.python == PYPY or "2.7" in session.python:
        local_deps += (DEPS["mock"],)
```

**Comment by dhermes** [comment]
@theacodes You're right. I even prefer what you did because `session.python == "2.7"` is superior to `"2.7" in session.python`. (Or `session.python in ("2.7", "2.7-32")` for those wild and crazy folks.)

**Comment by theacodes** [comment]
Oh cool, I thought I'd had a total brain fart. Glad you like the outcome.
